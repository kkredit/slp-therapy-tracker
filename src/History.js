import React from 'react';
import './History.css';

class History extends React.Component {
  render() {
    return (
      <h1>
        History page
      </h1>
    );
  }
};

export default History;
