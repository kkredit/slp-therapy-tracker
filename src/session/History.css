.historyRow {
  font-size: 10.5pt;
}

.modalHeader {
  /* text-align: center;
  font-weight: bold;
  font-size: 11pt; */
}

.modalStudentHeader {
  /* text-align: center;
  font-size: 10.5pt; */
}

.modalData .modalStudentData {
  /* text-align: center; */
}
