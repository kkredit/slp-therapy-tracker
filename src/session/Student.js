
export const STUDENTS_MIN = 1;
export const STUDENTS_MAX = 4;
export const GOALS_MIN = 1;
export const GOALS_MAX = 4;

export const ATTEMPT_SUCC = 0;
export const ATTEMPT_FAIL = 1;
export const ATTEMPT_CUED = 2;

export class Goal {
  constructor(number) {
    this.key = number;
    this.number = number;
    this.attempts = [];

    this.addAttempt = this.addAttempt.bind(this);
    this.addSuccess = this.addSuccess.bind(this);
    this.addFail = this.addFail.bind(this);
    this.addCued = this.addCued.bind(this);
  }

  addAttempt(type) {
    this.attempts.push(type);
  }

  addSuccess() {
    this.addAttempt(ATTEMPT_SUCC);
  }

  addFail() {
    this.addAttempt(ATTEMPT_FAIL);
  }

  addCued() {
    this.addAttempt(ATTEMPT_CUED);
  }
}

export class Student {
  constructor(number, updateCb) {
    this.key = number;
    this.number = number;
    this.goals = [new Goal(0)];
    this.updateCb = updateCb;

    this.addGoal = this.addGoal.bind(this);
    this.delGoal = this.delGoal.bind(this);
  }

  addGoal() {
    if (this.goals.length < GOALS_MAX) {
      this.goals.push(new Goal(this.goals.length));
      this.updateCb();
    }
  }

  delGoal() {
    if (this.goals.length > GOALS_MIN) {
      this.goals.pop();
      this.updateCb();
    }
  }
}
