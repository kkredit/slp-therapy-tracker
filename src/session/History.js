import React from 'react'
import axios from 'axios'
import Button from 'react-bootstrap/Button'
import ButtonGroup from 'react-bootstrap/ButtonGroup'
import NotificationAlert from '../helpers/NotificationAlert.js'

import './History.css'

const API_BASE = 'http://localhost:8000';

const HistoryListItem = (props) => {
  return (
    <tr>
      <td className="col-md-3">{props.time}</td>
      <td className="col-md-3">{props.provider}</td>
      <td className="col-md-3">{props.location}</td>
      <td className="col-md-3">
        <ButtonGroup aria-label="CRUD buttons">
          <Button variant="primary" size="sm">
            <i className="fa fa-eye"/> View
          </Button>
        </ButtonGroup>
      </td>
    </tr>
  );
}

const HistoryList = (props) => {
  const historyItems = props.history.map((history) => {
    return (
      <HistoryListItem
        time={history.time}
        provider={history.provider}
        location={history.location}
        id={history.id}
        key={history.id}
        onDelete={props.onDelete}
        onEdit={props.onEdit}
      />
    )
  });

  return (
    <div className="history-list">
      <table className="table table-hover">
        <thead>
          <tr>
            <th className="col-md-3">Time</th>
            <th className="col-md-3">Provider</th>
            <th className="col-md-3">Location</th>
            <th className="col-md-3">Actions</th>
          </tr>
        </thead>
        <tbody>
          {historyItems}
        </tbody>
      </table>
    </div>
  );
}

class History extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      historyItems: [],
      formMode: "new",
      history: {time:"", provider:"", location:"", id: "999999"},
      alertActive: false,
      alertVariant: '',
      alertText: ''
    };
    this.setAlert = this.setAlert.bind(this);
    this.clearAlert = this.clearAlert.bind(this);
    this.loadHistory = this.loadHistory.bind(this);
    this.removeHistory = this.removeHistory.bind(this);
  }

  render() {
    return (
      <div className="history">
        <h1> Session History </h1>
        <NotificationAlert active={this.state.alertActive}
                           variant={this.state.alertVariant}
                           text={this.state.alertText} />
        <HistoryList
          history={this.state.historyItems}
          onDelete={(id) => this.removeHistory(id)}
          onEdit={(mode,history) => this.updateForm(mode,history)}
        />
      </div>
    );
  }

  setAlert(variant, text) {
    this.setState({
      alertActive: true,
      alertVariant: '' + variant,
      alertText: text
    });
  }

  setAlertError(err, text) {
    this.setAlert('danger',
                  <React.Fragment>
                    <i className="fa fa-exclamation-triangle" aria-hidden="true"/><br />
                    {err.message}: {err.response.statusText}<br />
                    {text}
                  </React.Fragment>
                  );
  }

  clearAlert() {
    this.setState({
      alertActive: false,
      alertVariant: '',
      alertText: ''
    });
  }

  componentDidMount() {
    console.log('History mounted!')
    this.loadHistory();
  }

  loadHistory() {
    axios
      .get(`${API_BASE}/sessions.json`)
      .then(res => {
              this.setState({ history: res.data });
              this.clearAlert();
              console.log(`Data loaded! = ${this.state.history}`);
            })
      .catch(err => {
               console.log(err);
               this.setAlert('danger', err);
               this.setAlertError(err, "Could not load current history.");
             });
  }

  removeHistory(id) {
    let filteredArray = this.state.history.filter(item => item.id !== id)
    this.setState({history: filteredArray});
    axios
      .delete(`${API_BASE}/history/${id}.json`)
      .then(res => {
              console.log(`Record Deleted`);
              this.clearAlert();
            })
      .catch(err => {
               console.log(err);
               this.setAlertError(err, "Could not remove history.");
             });
  }
};

export default History;
