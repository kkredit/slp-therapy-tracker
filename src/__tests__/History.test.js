import React from 'react';
import ReactDOM from 'react-dom';
import History from '../session/History.js';
import { ATTEMPT_SUCC, ATTEMPT_CUED, ATTEMPT_FAIL } from '../session/Student.js'
import GetTestDb from '../__test_helpers__/SetupDb.js'
import { create } from 'react-test-renderer'
import { BrowserRouter } from 'react-router-dom'

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <BrowserRouter>
      <History db={GetTestDb()} />
    </BrowserRouter>
  , div);
  ReactDOM.unmountComponentAtNode(div);
});

describe('Snapshots',()=>{
  var tree;
  var instance;

  beforeEach(() => {
    let db = GetTestDb();

    tree = create(
      <BrowserRouter>
        <History db={db} />
      </BrowserRouter>
    )
    instance = tree.root.findByType(History)._fiber.stateNode;
  });

  afterEach(() => {
    tree = null;
    instance = null;
  });

  test('Initial setup', () => {
    expect(tree.toJSON()).toMatchSnapshot();
  })

  test('Add row with session data', () => {
    let updatedDb = GetTestDb();

    updatedDb.sessions = [
      {
        id: 1,
        time: new Date(),
        provider_id: updatedDb.providers[0].id,
        location_id: updatedDb.locations[0].id,
      }
    ];
    updatedDb.students = [
      {
        id: 1,
        number: 0,
        session_id: 1,
      }
    ];
    updatedDb.goals = [
      {
        id: 1,
        number: 0,
        student_id: 1,
      }
    ];
    updatedDb.attempts = [
      {
        goal_id: 1,
        status: ATTEMPT_SUCC,
      },
      {
        goal_id: 1,
        status: ATTEMPT_SUCC,
      },
      {
        goal_id: 1,
        status: ATTEMPT_CUED,
      },
      {
        goal_id: 1,
        status: ATTEMPT_FAIL,
      },
    ];

    // Must create new instance, since getDerivedStateFromProps() wipes out direct state db udpates
    tree = create(
      <BrowserRouter>
        <History db={updatedDb} />
      </BrowserRouter>
    )
    instance = tree.root.findByType(History)._fiber.stateNode;
    expect(tree.toJSON()).toMatchSnapshot();
  })
})
