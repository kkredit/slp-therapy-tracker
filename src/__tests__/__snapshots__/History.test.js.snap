// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots Add row with session data 1`] = `
<div
  className="history"
>
  <h1>
     Session History 
  </h1>
  <div
    className="history-list"
  >
    <table
      className="table table-hover"
    >
      <thead>
        <tr>
          <th
            className="col-md-2"
          >
            Date
          </th>
          <th
            className="col-md-2"
          >
            Time
          </th>
          <th
            className="col-md-3"
          >
            Provider
          </th>
          <th
            className="col-md-2"
          >
            Location
          </th>
          <th
            className="col-md-3"
          >
            Actions
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="historyRow"
        >
          <td
            className="col-md-2"
          >
            4/25/2019
          </td>
          <td
            className="col-md-2"
          >
            11:00:37 AM
          </td>
          <td
            className="col-md-3"
          >
            sally
            Â 
            foreman
          </td>
          <td
            className="col-md-2"
          >
            testroom1
          </td>
          <td
            className="col-md-3"
          >
            <div
              aria-label="CRUD buttons"
              className="btn-group"
              role="group"
            >
              <button
                className="btn btn-primary btn-sm"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                <i
                  className="fa fa-eye"
                />
                 View
              </button>
              <button
                className="btn btn-danger btn-sm"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                <i
                  className="fa fa-trash"
                />
                 Delete
              </button>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Snapshots Initial setup 1`] = `
<div
  className="history"
>
  <h1>
     Session History 
  </h1>
  <div
    className="history-list"
  >
    <table
      className="table table-hover"
    >
      <thead>
        <tr>
          <th
            className="col-md-2"
          >
            Date
          </th>
          <th
            className="col-md-2"
          >
            Time
          </th>
          <th
            className="col-md-3"
          >
            Provider
          </th>
          <th
            className="col-md-2"
          >
            Location
          </th>
          <th
            className="col-md-3"
          >
            Actions
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
